### PR TITLE
Pull Request: Feature/dangle

### PR DESCRIPTION
**Pull request by roboshoes**
This pull request adds a fuzzy way to calculate drift and rotation on crash. So far no physics engines have been used. Also `movable` gets an extra attribute called `torque`which represents the power released in ever curve segment. 

So far the dangeling of the car is not a hundred percent realistic but that might be a question of balancing. I though it might be nice to add it into the development branch for now to have it tested. 

_This entire feature is very easily removed, in case we hate it ..._

### CODE REVIEW

**Comment by stravid** [comment]
I suggest we don't merge this PR but keep it to discuss the current status of the feature.

My thoughts:
- I love the way the car is drifting out if you go to fast. Great work!
- The dangling in general is to fast / jumpy.

**Comment by roboshoes** [comment]
I fully agree. The dangling must be speed dependent (which it kind of is) but as soon as the car breaks down the dangling mussst be way less!

I keep working on it. Feedback appriciated!

**Comment by bmedicke** [comment]
moar dangle!

**Comment by roboshoes** [comment]
OK - I did a little fine tuning. Please, everybody check out the branch and do a little testing.

**Comment by stravid** [comment]
It's getting a lot better, nice job! But I still think we can do better, make it feel even more awesome :-)

**Comment by bmedicke** [comment]
We have to decide if we want to simulate a real car or a slot car. I think it looks fine for a slot car. Maybe the car dangles a bit too fast, but the general behavior (how far the back moves) is fine.

File: app/assets/javascripts/slotcars/play/views/car_view.js.coffee [hunk 1]
```diff
@@ -17,28 +17,25 @@
   puffStep: 0
 
   didInsertElement: ->
-    #@exhaust = helpers.graphic.Exhaust.create(@paper)
+    # @exhaust = helpers.graphic.Exhaust.create(@paper)
 
     # if @offset?
     #   (jQuery @$()).css 'top', @offset.top
     #   (jQuery @$()).css 'left', @offset.left
```
Comment (tombu) [inline app/assets/javascripts/slotcars/play/views/car_view.js.coffee:24]: I know this is nothing that belongs to your feature but as we don´t need `offset` any longer you could do me a favour an remove those 3 lines, couldn´t you? :wink:

**Comment by tombu** [comment]
Please, please, please write some comments! I can´t understand the (= your) logic in `moveTo()` ...

**Comment by roboshoes** [comment]
True - that can be removed just like the exhaust in general. it wouldn't work right now anyways. So it needs to be taken out.

File: app/assets/javascripts/slotcars/shared/lib/movable.js.coffee [hunk 2]
```diff
@@ -6,13 +6,78 @@ Vector = helpers.math.Vector
...
   position: x: 0, y: 0
+  bouncePosition: null
+  oldTailPosition: null
   rotation: 0
+  torque: 0
+  
+  # constants for bounce calculation
+  INERTIA: 0.8
+  OFFSET: 0.1
+  K: 0.16
```
Comment (DominikGuzei) [inline app/assets/javascripts/slotcars/shared/lib/movable.js.coffee:17]: what does K mean?

Please add comments to describe how these constants influence the behaviour of the car

File: app/assets/javascripts/slotcars/shared/lib/movable.js.coffee [hunk 3]
```diff
@@ -6,13 +6,78 @@ Vector = helpers.math.Vector
...
+  moveTo: (newPosition, tailPosition) ->
     previousPosition = @get 'position'
 
     @set 'position', newPosition
 
-    direction = Vector.create from: previousPosition, to: newPosition
-    @set 'rotation', direction.clockwiseAngle()
+    if @get 'isCrashing'
+      @set 'rotation', (@rotation + @torque)
+      @torque *= 0.94
```
Comment (DominikGuzei) [inline app/assets/javascripts/slotcars/shared/lib/movable.js.coffee:29]: no magic numbers -> move 0.94 to its own constant at the top of file with a descriptive name

File: app/assets/javascripts/slotcars/shared/lib/movable.js.coffee [hunk 4]
```diff
@@ -6,13 +6,78 @@ Vector = helpers.math.Vector
 (namespace 'slotcars.shared.lib').Movable = Ember.Mixin.create
```
Comment (DominikGuzei) [inline app/assets/javascripts/slotcars/shared/lib/movable.js.coffee:6]: Why is the code for dangle in Movable? Please move the whole dangle logic into an e.g `Dangleable` mixin to keep the file size small and the code testable!

**Comment by DominikGuzei** [comment]
Please move the whole dangle logic into its own mixin and add comments + comprehensive tests for the logic. It's ok to try things out and perfect everything without writing tests but before we merge that into the project we need good understandable code that is well tested and can be changed later on without shooting yourself ;-)

**Comment by roboshoes** [comment]
Unfortunatly you are completly wrong and I am right (as most of the time)!

There is a reason why this _logic_ is all in the movable. The dangeling only affects the rotation of the car which is definitely a part of movable. (Even by name: it moves the car). If you would have read my code closely you would have realized that I only work within this one function `moveTo`. Your argument to write it in form of inheritance (or for the sake of that as a mixin) and override `moveTo` would be completely senseless and collide with almost every coding law because it would create pointless unused code that **MIGHT ONE DAY IN 17 YEARS** be used by somebody who forks the project and creates another game.
In our specific case. I rewrote the calculation of the angle of the car (which by the way also fixes the jumpstart mess). I can not reuse the old `moveTo` function so overriding it would just through it away.
Also - the only thing, that is public, that I added is the `torque` which is tested by the way. You all would agree that you would not test lines within a function, or would you?

So in short I believe you can either change `Movable` or create a `Dangable`and throw `Movable` away which in other words would mean renaming it.

PS: I will rename most of the variables.
